testnet/src/node.rs

```diff
@@ -338,6 +492,114 @@ impl Node {
+
+        let stacks_tip = match chain_state.get_stacks_chain_tip().unwrap() {
+            Some(x) => x,
+            None => {
+                warn!("Could not mine new tenure, since no chain tip known.");
+                return None
+            }
+        };
+
+        let stacks_tip_header = match StacksChainState::get_anchored_block_header_info(
+            &chain_state.headers_db, &stacks_tip.burn_header_hash, &stacks_tip.anchored_block_hash).unwrap() {
+            Some(x) => x,
+            None => {
+                warn!("Could not mine new tenure, since could not find header for known chain tip.");
+                return None
+            }
+        };
+
+
+        debug!("Mining tenure on burn_block: {}, stacks tip burn_header_hash: {}",
+              &burn_block.block_snapshot.burn_header_hash,
+              &stacks_tip.burn_header_hash);
+
+        // Generates a proof out of the sortition hash provided in the params.
+        let vrf_proof = keychain.generate_proof(
+            &registered_key.vrf_public_key, 
+            burn_block.block_snapshot.sortition_hash.as_bytes()).unwrap();
+
+        debug!("Generated VRF Proof: {} over {} with key {}",
+              vrf_proof.to_hex(),
+              &burn_block.block_snapshot.sortition_hash,
+              &registered_key.vrf_public_key.to_hex());
+
+        // Generates a new secret key for signing the trail of microblocks
+        // of the upcoming tenure.
+        let microblock_secret_key = keychain.rotate_microblock_keypair();
+
+        let burn_header_to_poll = &stacks_tip.burn_header_hash;
+        let block_hash_to_poll = &stacks_tip.anchored_block_hash;
+
+        let ratio = StacksWorkScore {
+            burn: burn_block.block_snapshot.total_burn,
+            work: 1 + stacks_tip_header.anchored_header.total_work.work,
+        };
+
+        // relayer is always building a non-genesis block.
+        let mut block_builder = StacksBlockBuilder::from_parent(
+            1, 
+            &stacks_tip_header,
+            &ratio,
+            &vrf_proof, 
+            &microblock_secret_key);
+
+        let mut clarity_tx = block_builder.epoch_begin(chain_state).unwrap();
+        
+
+        // make a coinbase
+
+        let principal = keychain.origin_address().unwrap().into();
+        let nonce = StacksChainState::get_account(&mut clarity_tx, &principal).nonce;
+
+        let coinbase_tx = Node::inner_generate_coinbase_tx(keychain, nonce);
+        block_builder.try_mine_tx(&mut clarity_tx, &coinbase_tx);
+
+        let txs = mem_pool.poll(burn_header_to_poll, &block_hash_to_poll);
+
+        for tx in txs {
+            let res = block_builder
+                .try_mine_tx(&mut clarity_tx, &tx);
+            match res {
+                Err(e) => error!("Failed mining transaction - {}", e),
+                Ok(_) => {},
+            };
+        }
```

Comment:
the new `Tenure.run` from master had a sleeping behavior, in case of the mempool is empty / almost empty. This was not a desired behavior?

Reply:
Since this relayer_run_tenure function shares duties with handling the NetworkHandle events from the p2p network, we don't want this function to sleep -- otherwise we could sleep() here, while network events (like new mempool txs) are waiting to be processed. I think that we might want to support something like the wait-until-deadline behavior, but it would probably need to be triggered via a timer thread that sends a RelayerDirective.